python/ql/src/experimental/Security/CWE-287-ConstantSecretKey/ConstantSecretKey.ql

```diff
@@ -0,0 +1,219 @@
+                .getReturn(),
+            // legacy API https://docs.python.org/3/library/configparser.html#legacy-api-examples
+            API::moduleImport("configparser")
+                .getMember(["ConfigParser", "RawConfigParser"])
+                .getReturn()
+                .getMember("get")
+                .getReturn()
+          ] and
+        source = cn.asSource()
+      )
+      or
+      exists(API::CallNode cn |
+        cn =
+          [
+            API::moduleImport("os").getMember("getenv").getACall(),
+            API::moduleImport("os").getMember("environ").getMember("get").getACall()
+          ] and
+        (
+          // this can be ideal if we assume that best security practice is that
+          // we don't get SECRET_KEY from env and we always assign a secure generated random string to it
+          cn.getNumArgument() = 1
+          or
+          cn.getNumArgument() = 2 and
+          DataFlow::localFlow(any(DataFlow::Node n | n.asExpr().isConstant()), cn.getArg(1))
+        ) and
+        source.asExpr() = cn.asExpr()
+      )
+      or
+      exists(DataFlow::LocalSourceNode lsn |
+        lsn = API::moduleImport("os").getMember("environ").getASubscript().asSource() and
+        source.asExpr() = lsn.asExpr()
+      )
+    ) and
+    not source.getScope().getLocation().getFile().inStdlib()
+  }
+
+  /**
+   * Sinks are one of the following kinds, some of them are directly connected to a flask Instance like
+   * ```python
+   *    app.config['SECRET_KEY'] = 'CHANGEME1'
+   *    app.secret_key = 'CHANGEME2'
+   *    app.config.update(SECRET_KEY="CHANGEME3")
+   *    app.config.from_mapping(SECRET_KEY="CHANGEME4")
+   * ```
+   * other Sinks are SECRET_KEY Constants Variables that are defined in seperate files or a class in those files like:
+   * ```python
+   *    app.config.from_pyfile("config.py")
+   *    app.config.from_object('config.Config')
+   *```
+   * we find these files with `FromObjectFileName` DataFlow Configuration
+   * note that "JWT_SECRET_KEY" is same as "SECRET_KEY" but it is belong to popular flask-jwt-extended library
+   */
+  predicate isSink(DataFlow::Node sink) {
+    (
+      exists(API::Node n |
+        n = flaskInstance() and
+        flask_sessionSanitizer(n.getReturn().asSource())
+      |
+        sink =
+          [
+            n.getReturn().getAMember().getSubscript(["SECRET_KEY", "JWT_SECRET_KEY"]).asSink(),
+            n.getReturn().getMember(["SECRET_KEY", "JWT_SECRET_KEY"]).asSink(),
+            n.getReturn()
+                .getMember("config")
+                .getMember(["update", "from_mapping"])
+                .getACall()
+                .getArgByName(["SECRET_KEY", "JWT_SECRET_KEY"])
+          ]
+      )
+      or
+      // this query checks for Django SecretKey too
+      if exists(API::moduleImport("django"))
+      then
+        exists(AssignStmt e | e.getTarget(0).toString() = "SECRET_KEY" |
```

Comment:
No use of `.toString()` in production code.

You probably want something like `.(Name).getId() = "SECRET_KEY"`